src/blockchain/process-block.ts

```diff
@@ -0,0 +1,47 @@
+import Augur from "augur.js";
+import * as Knex from "knex";
+import { logError } from "../utils/log-error";
+import { BlocksRow } from "../types";
+
+export function processBlock(db: Knex, augur: Augur, blockNumberString: string): void {
+  augur.rpc.eth.getBlockByNumber([blockNumberString, false], (block: any): void => {
+    if (!block || block.error || !block.timestamp) return logError(new Error(JSON.stringify(block)));
+    const blockNumber = parseInt(blockNumberString, 16);
+    const timestamp = parseInt(block.timestamp, 16);
+    console.log("new block:", blockNumber, timestamp);
+    db.transaction((trx: Knex.Transaction): void => {
+      trx("blocks").where({ blockNumber }).asCallback((err: Error|null, blocksRows?: Array<BlocksRow>): void => {
+        if (err) return logError(err);
+        let query: Knex.QueryBuilder;
+        if (!blocksRows || !blocksRows.length) {
+          query = db.transacting(trx).insert({ blockNumber, timestamp }).into("blocks");
+        } else {
+          query = db("blocks").transacting(trx).where({ blockNumber }).update({ timestamp });
```

Comment:
Is this logic required any more? Seems like if we are appropriately removing blocks, there's no need for an UPSERT?

Reply:
That's a good point -- I think this is needed until Paul's fixes to initial sync (using blockstream etc.) go in because the way things are now, we can technically get duplicate block notifications (from initial sync, then again from block listener).